### ISSUE TITLE
MSE Loss Implementation

### ISSUE THREAD

**Issue by KsenijaS** [pytorch/pytorch#33787]
During the implementation of ONNX export of mse loss function I encountered a problem with broadcastable tensors (not supported in ONNX), and I have a couple of questions about certain implementation details of mse loss in Pytorch.

1. Documentation of MSELoss states that input and target tensors should be of the same shape:
https://pytorch.org/docs/stable/nn.html#mseloss, so I don't understand why are broadcastable tensors used in mse_loss function: https://github.com/pytorch/pytorch/blob/a9cef05f5d6ca2a74f993feb191531a03e445d40/torch/nn/functional.py#L2523

2. When required_grad is set to True for target tensor, tensors don't need to be broadcastable: https://github.com/pytorch/pytorch/blob/a9cef05f5d6ca2a74f993feb191531a03e445d40/torch/nn/functional.py#L2518. Why is the condition of requires grad only applied for target and not for input as well?

@houseroad

**Comment by houseroad** [comment]
Actually, this is not ONNX specific question. It's a general question about why we implement this MSELoss in such way. Is there any problem with C++ backward function so we try to avoid it?

cc: @albanD

**Comment by albanD** [comment]
Hi,

> Documentation of MSELoss states that input and target tensors should be of the same shape:

Yes the doc states this, but the implementation has always been doing broadcasting. So for BC-breaking reasons, we need to keep supporting broadcasting. And added a warning when the user breaks that assumption.

> When required_grad is set to True for target tensor, tensors don't need to be broadcastable:

They actually do, just that the broadcasting logic is done by the subtraction `input - target`.

> Why is the condition of requires grad only applied for target and not for input as well?

The efficient cpp implementation has the limitation that it cannot compute gradients for the target. So when only use it when we can, and in other cases, we fall back on a less efficient implementation that is more general.

Does that answer all your questions?

**Comment by KsenijaS** [comment]
Hi Alban,

Thank you for answering my questions, it is very helpful. I have one additional question though:
mse_loss operator is not used if target.requires_grad is set to true, and is instead calculated using other operators. My understanding from previous PRs to mse_loss is that there is a bug in mse_loss function and that's why things are done this way.  So my question is, are there plans to fix that bug and use mse_loss operator always instead of calculating mse loss?
The reason I'm asking is that for ONNX and  ORT training especially is important to have one  mse_loss operator instead  of a couple of operators that calucate mse loss.

**Comment by albanD** [comment]
I wouldn't call it a bug for the mse_loss cpp/cuda operator. It was designed this way.

But a feature request to add support for backprop through the target in the cpp operator would be welcome accepted.
But I don't think we have such feature request so I don't think anyone is working on it.
